Bootstrap entitlements for testing

prdoyle opened the pull request:
Add entitlement enforcement during ordinary unit tests.

This does not yet cover tests that run ES nodes; only ordinary unit tests.

See ES-11597.

prdoyle committed:
Ignore server locations whose representative class isn't loaded

prdoyle committed:
Fix ExceptionSerializationTests to use getCodeSource instead of getResource.

Using getResource makes this sensitive to unrelated classpath entries,
such as the entitlement bridge library, that get prepended to the classpath.

prdoyle committed:
FIx logging tests to use org.elasticsearch.index instead of root logger.

Using the root logger makes this sensitive to unrelated
logging, such as from the entitlement library.

prdoyle committed:
Fix entitlement error message by stashing the module name in ModuleEntitlements.

Taking the actual module name from the class doesn't work in tests,
where those classes are loaded from the classpath and so their module
info is misleading.

prdoyle committed:
Partial initial implementation

prdoyle committed:
Add the classpath to the source path list for every plugin

prdoyle committed:
System properties: testOnlyClasspath and enableForTests

prdoyle committed:
Trivially allow some packages

prdoyle committed:
DEBUG: use TreeMap in TestScopeResolver for readability

prdoyle committed:
Special case bouncycastle for security plugin

prdoyle committed:
Add CONFIG to TestPathLookup

prdoyle committed:
Add @WithoutEntitlements to tests that run ES nodes

prdoyle committed:
Set es.entitlement.enableForTests for all libs

prdoyle committed:
Use @WithoutEntitlements on ingest plugin tests

prdoyle committed:
Substitute ALL-UNNAMED for module name in non-modular plugins

prdoyle committed:
Add missing entitlements found by unit tests

prdoyle committed:
Comment in TestScopeResolver

prdoyle committed:
Properly compute bridge jar location for patch-module

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

ldematte reviewed (commented):
I think this looks good; I've skipped `configureEntitlements` as I have far from the necessary gradle knowledge.
My main concern is with `TestPolicyManager`/`PolicyManager` (see comments)

prdoyle commented on server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java:
```diff
@@ -267,11 +267,18 @@ private static void initPhase2(Bootstrap bootstrap) throws IOException {
+    /**
+     * @throws IllegalStateException if entitlements aren't functioning properly.
+     */
+    static void entitlementSelfTest() {
```
Yes that's right. The denied method call itself needs to be in a separate class for reasons described elsewhere in the same file. That class is private. So I had the choice to either (1) expose the class itself, or (2) expose a method that calls the private class. The latter seemed less "leaky".

prdoyle commented on test/framework/src/main/java/org/elasticsearch/entitlement/runtime/policy/TestPolicyManager.java:
```diff
@@ -10,30 +10,62 @@
+    boolean isTriviallyAllowingTestCode;
+
+    /**
+     * We don't have modules in tests, so we can't use the inherited map of entitlements per module.
```
Composition > inheritance of abstract classes > inheritance of concrete classes. 😂 

That's a great suggestion.

ldematte replied:
Technically we would just shuffle code around, moving some things from `PolicyManager` to `AbstractPolicyManager`, but...

> avoid adding complexity to production code solely for the sake of testing

Yes, that's true and it's always a good principle; however we do this all the time in ES :) not a justification, just a point. It's true that we would add some complexity to production code, but we would separate a concern; the concrete PolicyManager would be effectively a `ModuleBasedPolicyManager`, if you think about it: I believe the things that will be moved there that we don't want/need to share with the TestPolicyManager are those related to modules.
And I'd weigh in the "clarity" (or confusion) of having unused stuff in TestPolicyManager, especially if someone is debugging a entitlement exception in a unit test.

We can maybe try it in a separate PR (even a follow-up if we are tight in time to close this)

prdoyle replied:
Ok that all makes sense. Let's do this in another PR.

prdoyle replied:
BTW I love the name `ModuleBasedPolicyManager`.

prdoyle commented on x-pack/plugin/core/src/main/plugin-metadata/entitlement-policy.yaml:
```diff
@@ -18,6 +18,7 @@ org.apache.httpcomponents.httpasyncclient:
   - manage_threads
 unboundid.ldapsdk:
   - set_https_connection_properties # TODO: review if we need this once we have proper test coverage
+  - inbound_network # For com.unboundid.ldap.listener.LDAPListener
```
I'm not sure. It was definitely exercised by the unit test.

I wonder who would know. Does our team own x-pack-core?

ldematte replied:
It's kind of unrelated, so if a test hits it it's worth adding it. But I think this is an anomaly, and needs to be investigate (later -- I'm completely fine adding this now to make progress!)
By whoever "owns" x-pack-code, which is going to be really tricky. I _guess_ since this is related to authN/authZ this should be security?

prdoyle committed:
Call out nonServerLibs

prdoyle committed:
Don't build two TestPathLookups

prdoyle committed:
More comments for meta-tests

prdoyle committed:
Remove redundant dependencies for bridgeJarConfig.

These are alread set in ElasticsearchJavaBasePlugin.

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Add bridge+agent dependencies only if those exist.

For serverless, those project dependencies don't exist, and we'll need
to add the dependencies differently, using Maven coordinates.

Commit:
[CI] Auto commit changes from spotless

prdoyle committed:
Pass testOnlyPath in environment instead of command line.

It's typically a very very long string, which made Windows angry.

Commit:
[CI] Auto commit changes from spotless

prdoyle committed:
Split testOnlyPathString at File.pathSeparator

prdoyle committed:
Use doFirst to delay setting testOnlyPath env var

prdoyle committed:
Trivially allow jimfs (??)

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Don't enforce entitlements on internalClusterTest for now

prdoyle commented on build-tools/src/main/java/org/elasticsearch/gradle/test/TestBuildInfoPlugin.java:
```diff
@@ -53,5 +56,11 @@ public void apply(Project project) {
         });
+
+        project.getTasks().withType(Test.class).configureEach(test -> {
+            if (List.of("test", "internalClusterTest").contains(test.getName())) {
```
Update: I've removed `internalClusterTest` because some serverless tests were failing. We can re-add it in follow-on work.

ldematte committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Replace forbidden APIs

prdoyle committed:
Match testOnlyClasspath using URI instead of String.

We already get the "needle" in the form of a URI, so this skips
a step, and has the benefit of also working on Windows.

Commit:
[CI] Auto commit changes from spotless

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
More forbidden APIs

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

ldematte reviewed (approved):
As agreed, we will defer the PolicyManager/TestPolicyManager refactoring to a follow-up PR (this one is already big enough).
My advice is to get a second pair of eyes on the gradle stuff, but provided CI pass this LGTM.

prdoyle committed:
Disable configuration cache for LegacyYamlRestTestPluginFuncTest

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into bootstrap-entitlements-for-testing

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:
```diff
@@ -196,6 +197,49 @@ private static void configureNativeLibraryPath(Project project) {
+        FileCollection agentFiles = agentJarConfig;
+
+        Configuration bridgeJarConfig = project.getConfigurations().create("entitlementBridgeJar");
+        Project bridge = project.findProject(":libs:entitlement:bridge");
```
our binary plugin implementations should not know anything of our gradle project structure. Any dependency declarations should live in the build scripts or in a convention plugin. None of this logic IMO belongs in the java base plugin as the java base plugin does only deal with java base buidling and also should not handle test configurations.

breskeby replied:
I think all of this here should for now at least be be part of the ElasticsearchTestBase plugin as its all about dealing with the test setup. Ultimately we want those project dependencies to be moved out from here into a convention plugin. We (I can take care of that in a folllow up)

Regarding

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:
```diff
@@ -196,6 +197,49 @@ private static void configureNativeLibraryPath(Project project) {
+        Project bridge = project.findProject(":libs:entitlement:bridge");
+        if (bridge != null) {
+            bridgeJarConfig.defaultDependencies(deps -> {
+                deps.add(project.getDependencies().project(Map.of("path", ":libs:entitlement:bridge", "configuration", "default")));
```
```suggestion
                deps.add(project.getDependencies().project(Map.of("path", ":libs:entitlement:bridge")));
```

Relying on the default configuration when declaring a project dependency should be avoided and not necessary here

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -173,6 +175,16 @@ public void execute(Task t) {
+                FileCollection mainRuntime = mainSourceSet.getRuntimeClasspath();
+                FileCollection testRuntime = testSourceSet.getRuntimeClasspath();
+                FileCollection testOnlyFiles = testRuntime.minus(mainRuntime);
+                test.doFirst(task -> test.environment("es.entitlement.testOnlyPath", testOnlyFiles.getAsPath()));
```
```suggestion
                Provider<String> testOnlyPathProvider = getProviderFactory().provider(() -> testOnlyFiles.getAsPath());
                test.environment("es.entitlement.testOnlyPath", testOnlyPathProvider);
```

prdoyle replied:
When I make this change, I get the following error:

```
Error while saving task graph
> Configuration cache state could not be cached: field `environment` of `org.gradle.process.internal.DefaultJavaForkOptions_Decorated` bean found in field `forkOptions` of task `:plugins:discovery-ec2:test` of type `org.gradle.api.tasks.testing.Test`: error writing value of type 'java.util.HashMap'
   > Could not resolve all files for configuration ':plugins:discovery-ec2:testRuntimeClasspath'.
      > Failed to transform ec2-imds-fixture-9.1.0-SNAPSHOT.jar (project :test:fixtures:ec2-imds-fixture) to match attributes {artifactType=jar, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.jvm.version=21, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime, patched=true}.
         > Execution failed for Awsv2ClassPatcher: /Users/prdoyle/IdeaProjects/elasticsearch/test/fixtures/ec2-imds-fixture/build/distributions/ec2-imds-fixture-9.1.0-SNAPSHOT.jar.
            > Transform output /Users/prdoyle/IdeaProjects/elasticsearch/test/fixtures/ec2-imds-fixture/build/distributions/ec2-imds-fixture-9.1.0-SNAPSHOT.jar must exist.
> Configuration cache state could not be cached: field `environment` of `org.gradle.process.internal.DefaultJavaForkOptions_Decorated` bean found in field `forkOptions` of task `:x-pack:extras:plugins:microsoft-graph-authz:test` of type `org.gradle.api.tasks.testing.Test`: error writing value of type 'java.util.HashMap'
   > Could not resolve all files for configuration ':x-pack:extras:plugins:microsoft-graph-authz:testRuntimeClasspath'.
      > Failed to transform x-pack-core-test-artifacts-9.1.0-SNAPSHOT.jar (project :x-pack:plugin:core) to match attributes {artifactType=jar, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.jvm.version=21, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime, patched=true}.
         > Execution failed for AzureCoreClassPatcher: /Users/prdoyle/IdeaProjects/elasticsearch/x-pack/plugin/core/build/distributions/x-pack-core-test-artifacts-9.1.0-SNAPSHOT.jar.
            > Transform output /Users/prdoyle/IdeaProjects/elasticsearch/x-pack/plugin/core/build/distributions/x-pack-core-test-artifacts-9.1.0-SNAPSHOT.jar must exist.
```

breskeby replied:
keep it for now then. damn you gradle and our damn patch artifact transforms

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -173,6 +175,16 @@ public void execute(Task t) {
+            SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+            SourceSet mainSourceSet = sourceSets.findByName(SourceSet.MAIN_SOURCE_SET_NAME);
+            SourceSet testSourceSet = sourceSets.findByName(SourceSet.TEST_SOURCE_SET_NAME);
+            if (mainSourceSet != null && testSourceSet != null) {
```
Thinking about this more you might wanna check for the test task name earlier and don't deal with those sourceSets dependencies and in the first place if its not `test`

prdoyle replied:
I think I need to deal with `internalClusterTest` too, eventually, once we get to `ES-12042`.

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -205,19 +217,18 @@ public void execute(Task t) {
                     // Add the shadow JAR artifact itself
                     FileCollection shadowJar = project.files(project.getTasks().named("shadowJar"));
-                    FileCollection testRuntime = sourceSets.getByName(SourceSet.TEST_SOURCE_SET_NAME).getRuntimeClasspath();
+                    FileCollection mainRuntime = requireNonNull(mainSourceSet).getRuntimeClasspath();
```
why would that be required now?

prdoyle replied:
Oh I changed this and my intention was to change it back, but I think I flubbed it. Let me try again.

prdoyle replied:
This code is supposed to be equivalent to what it was before, except that now I define `mainSourceSet` and `testSourceSet` already higher up in the file.

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -235,13 +246,28 @@ private void configureImmutableCollectionsPatch(Project project) {
             test.getInputs().files(patchedFileCollection);
             test.systemProperty("tests.hackImmutableCollections", "true");
+
+            Configuration bridgeJarConfig = configurations.findByName("entitlementBridgeJar");
```
why is half of this living here and half of the stuff in ElasticsearchJavaBase plugin?

prdoyle replied:
I'm moving it all to ElasticsearchTestBasePlugin.

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -235,13 +246,28 @@ private void configureImmutableCollectionsPatch(Project project) {
         var deps = project.getDependencies();
         deps.add(configurationName, deps.project(Map.of("path", patchProject, "configuration", "patch")));
+
+        // If ElasticsearchJavaBasePlugin has specified a dependency on the entitlement bridge jar,
```
Again, just move what you did in ElasticsearchJavaBasePlugin over here.

breskeby commented on libs/build.gradle:
```diff
@@ -45,4 +45,16 @@ configure(childProjects.values()) {
+  // Omit oddball libraries that aren't in server.
+  def nonServerLibs = Set.of('plugin-scanner')
+  if (false == nonServerLibs.contains(project.name)) {
+    project.getTasks().withType(Test.class).configureEach(test -> {
```
project.getTasks().withType(Test.class).named('test').configure {

breskeby replied:
```suggestion
    project.getTasks().withType(Test.class).named("test").configure { test -> {
```

breskeby commented on libs/build.gradle:
```diff
@@ -45,4 +45,16 @@ configure(childProjects.values()) {
     apply plugin: 'elasticsearch.build'
   }
+
+  // This is for any code potentially included in the server at runtime.
```
Can we just add this to the plugin-scanner project for now. In general we want to move more configuration where it belongs. adding code that might be needed in the future in sometime just adds clutter for now.

prdoyle replied:
It needs to apply to every lib _except_ plugin-scanner.

prdoyle replied:
(Really, plugin-scanner shouldn't be in `libs` in the first place.)

breskeby reviewed (changes requested):
I added some remarks regarding the build changes that I think we need to address.

prdoyle committed:
Strip carriage-return characters in expected output for ReleaseNotesGeneratorTest.

The template generator also strips these, so we need to do so to make this pass
on Windows.

Note that we use replace("\r", "") where the template generator uses
replace("\\r", ""). The latter didn't work for me when I tried it on Windows,
for reasons I'm not aware of.

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Move configureEntitlements to ElasticsearchTestBasePlugin as-is

prdoyle commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:
```diff
@@ -196,6 +197,49 @@ private static void configureNativeLibraryPath(Project project) {
+        }
+        FileCollection bridgeFiles = bridgeJarConfig;
+
+        project.getTasks().withType(Test.class).configureEach(test -> {
```
There is no need to apply this configuration to tests that run ES nodes in separate processes, because those already work. I only need to configure tests that use the JUnit launcher to run ES code, including internal cluster tests.

prdoyle commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -37,6 +38,7 @@
 
 import javax.inject.Inject;
 
+import static java.util.Objects.requireNonNull;
```
Ok, I'll remove it. I added it to silence warnings in my IDE due to using `findByName` instead of `getByName` higher up in the file.

prdoyle commented on build-tools/src/main/java/org/elasticsearch/gradle/test/TestBuildInfoPlugin.java:
```diff
@@ -53,5 +56,11 @@ public void apply(Project project) {
             task.into("META-INF", copy -> copy.from(testBuildInfoTask));
         });
+
+        project.getTasks().withType(Test.class).configureEach(test -> {
```
That `contains` logic is anticipating ES-12042, where we're planning to add `internalClusterTest`.

prdoyle committed:
Use matching instead of if

prdoyle committed:
Remove requireNonNull

prdoyle committed:
Remove default configuration

prdoyle committed:
Set inputs instead of dependencies

prdoyle committed:
Use test.systemProperty

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into bootstrap-entitlements-for-testing

mosche commented on test/framework/src/main/java/org/elasticsearch/entitlement/runtime/policy/TestPolicyManager.java:
```diff
@@ -44,16 +78,91 @@ protected boolean isTrustedSystemClass(Class<?> requestingClass) {
+
+    @Override
+    protected ModuleEntitlements getEntitlements(Class<?> requestingClass) {
+        return classEntitlementsMap.computeIfAbsent(requestingClass, c -> computeEntitlements(requestingClass));
```
```suggestion
        return classEntitlementsMap.computeIfAbsent(requestingClass, this::computeEntitlements);
```

mosche commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -205,24 +216,24 @@ public void execute(Task t) {
+        configureEntitlements(project);
     }
 
     private void configureImmutableCollectionsPatch(Project project) {
```
I suggest renaming this, this isn't limited to immutable collections anymore

prdoyle replied:
For context: it is illegal to pass multiple `--patch-module` options for the same module (in this case, `java.base`). All patching for a given module needs to be smushed into the same command line option.

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

breskeby commented on libs/build.gradle:
```diff
@@ -45,4 +45,14 @@ configure(childProjects.values()) {
+
+  // This is for any code potentially included in the server at runtime.
+  // Omit oddball libraries that aren't in server.
+  def nonServerLibs = Set.of('plugin-scanner')
```
```suggestion
  def nonServerLibs = ['plugin-scanner']
```

mosche commented:
@prdoyle I noticed that file entitlement checks likely don't work at all in tests as they are using `org.apache.lucene.tests.mockfile.FilterFileSystem` rather than the default file system. In policy checker we're allowing everything that's not using the default file system

mosche commented on test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java:
```diff
@@ -71,6 +73,21 @@ public class BootstrapForTesting {
+
+        // Fire up entitlements
+        try {
+            TestEntitlementBootstrap.bootstrap(javaTmpDir, maybePath(System.getProperty("tests.config")));
```
I can only find a single reference of `tests.config` where this references elasticsearch.yml, though bootstrap expects a dir.
```
mvn -Dtests.gce=true -Dtests.config=/path/to/config/file/elasticsearch.yml clean test
```

prdoyle committed:
Respond to PR comments

prdoyle committed:
Disable entitlement enforcement for ScopedSettingsTests.

This test works by altering the logging on the root logger.
With entitlements enabled, that will cause additional log statements to appear,
which interferes with the test.

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into bootstrap-entitlements-for-testing

mosche commented:
@breskeby could you have another look for delivery? I'm looking into the bwc failures... would be good to wrap this up as it's blocking further work.

mosche commented on build-tools/src/main/java/org/elasticsearch/gradle/test/TestBuildInfoPlugin.java:
```diff
@@ -53,5 +56,9 @@ public void apply(Project project) {
         });
+
+        project.getTasks().withType(Test.class).matching(test -> List.of("test").contains(test.getName())).configureEach(test -> {
+            test.systemProperty("es.entitlement.enableForTests", "true");
```
No, everything fails for serverless if enabled... :(

breskeby replied:
@mosche I _think_ one (not the prettiest way) to approch this is to wrap this in 
```
if(project.getGradle().getParent() == null) {
                
 }
```
for now

prdoyle replied:
Actually it seems to be many tests, which makes sense.

I wonder what it would take to get it working in serverless. 🤔

mosche replied:
yes, good question... it could also be `configureJavaBasePatch`

prdoyle replied:
Awesome. I'm also going to rename it again to `configureJavaBaseModuleOptions` since it also does `--add-opens`.

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -235,16 +246,74 @@ private void configureImmutableCollectionsPatch(Project project) {
         });
     }
+
+    private static void configureEntitlements(Project project) {
```
Can you briefly add a javadoc here explaining what this is actually doing?

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -235,16 +246,74 @@ private void configureImmutableCollectionsPatch(Project project) {
             test.getInputs().files(patchedFileCollection);
             test.systemProperty("tests.hackImmutableCollections", "true");
+
+            Configuration bridgeJarConfig = configurations.findByName("entitlementBridgeJar");
```
nit: can we just name the configuration `entitlementBridge`

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchTestBasePlugin.java:
```diff
@@ -235,16 +246,74 @@ private void configureImmutableCollectionsPatch(Project project) {
     }
+
+    private static void configureEntitlements(Project project) {
+        Configuration agentJarConfig = project.getConfigurations().create("entitlementAgentJar");
```
nit: can we just name the configuration `entitlementAgent`

prdoyle committed:
Address PR comments

prdoyle committed:
Moritz's configureJavaBaseModuleOptions

prdoyle committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle committed:
Allow for entitlements not yet enforced in serverless

mosche committed:
fix entitlementBridge config after rename

mosche committed:
drop empty file collections

mosche committed:
Merge branch 'main' into bootstrap-entitlements-for-testing

prdoyle commented:
@mosche - if we're backporting to 9.0, should we also backport to 8.18?

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into bootstrap-entitlements-for-testing

prdoyle committed:
Remove workaround in LegacyYamlRestTestPluginFuncTest

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.19  | Commit could not be cherrypicked due to conflicts |
| ❌ |  9.0  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.18  | Commit could not be cherrypicked due to conflicts |
| ✅ |  [9.1](https://github.com/elastic/elasticsearch/pull/130483)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/130483">](https://github.com/elastic/elasticsearch/pull/130483) |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 129268`
